src/gui/passkeys/PasskeyImporter.cpp

```diff
@@ -0,0 +1,139 @@
+/*
+ *  Copyright (C) 2023 KeePassXC Team <team@keepassxc.org>
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "PasskeyImporter.h"
+#include "PasskeyImportDialog.h"
+#include "browser/BrowserPasskeys.h"
+#include "core/Entry.h"
+#include "core/Group.h"
+#include "gui/FileDialog.h"
+#include "gui/MessageBox.h"
+#include <QFileInfo>
+#include <QTextStream>
+#include <QUuid>
+
+static const QString IMPORTED_PASSKEYS_GROUP = QStringLiteral("Imported Passkeys");
+
+void PasskeyImporter::importPasskey(QSharedPointer<Database>& database)
+{
+    Q_UNUSED(database)
+    auto filter = QString("%1 (*.passkey);;%2 (*)").arg(tr("Passkey file"), tr("All files"));
+    auto fileName =
+        fileDialog()->getOpenFileName(nullptr, tr("Open Passkey file"), FileDialog::getLastDir("passkey"), filter);
+    if (fileName.isEmpty()) {
+        return;
+    }
+
+    FileDialog::saveLastDir("passkey", fileName, true);
+
+    QFile file(fileName);
+    if (!file.open(QIODevice::ReadOnly | QIODevice::Text)) {
+        MessageBox::information(
+            nullptr, tr("Cannot open file"), tr("Cannot open file \"%1\" for reading.").arg(fileName));
+        return;
+    }
+
+    importSelectedFile(file, database);
+}
+
+void PasskeyImporter::importSelectedFile(QFile& file, QSharedPointer<Database>& database)
+{
+    QTextStream fileStream(&file);
+    const auto relyingParty = fileStream.readLine();
+    const auto username = fileStream.readLine();
+    const auto password = fileStream.readLine();
+    const auto userHandle = fileStream.readLine();
+
+    QString privateKey;
+    while (!fileStream.atEnd()) {
+        privateKey.append(fileStream.readLine() + "\n");
+    }
+
+    if (relyingParty.isEmpty() || username.isEmpty() || password.isEmpty() || userHandle.isEmpty()
+        || privateKey.isEmpty()) {
+        MessageBox::information(nullptr,
+                                tr("Cannot import Passkey"),
+                                tr("Cannot import Passkey file \"%1\". Data is missing.").arg(file.fileName()));
+        return;
+    } else if (!privateKey.startsWith("-----BEGIN PRIVATE KEY-----")
+               || !privateKey.endsWith("-----END PRIVATE KEY-----\n")) {
```

Comment:
This might get tripped up if there is no ending newline in the file which isn't required. Might be best to use a REGEX here.

Reply:
Decided to go with an easy solution using `.trimmed()`.